services_app/src/main/java/org/opendatakit/services/preferences/fragments/ServerSettingsFragment.java

```diff
@@ -273,25 +305,172 @@ public boolean onPreferenceChange(Preference preference, Object newValue) {
+        } else {
+          parseQrCodeResult(result.getContents());
+          Log.i("QR code:",result.getContents());
+        }
+      }
+    super.onActivityResult(requestCode, resultCode, data);
 
 }
+
+  private void parseQrCodeResult(String contents) {
+    String TAG =  "pasreQrCodeResult";
+    String toastString = "";
+    try {
+     JSONObject mainObject = new JSONObject(contents);
+      try{
+        String url = (String)mainObject.get("url");
+        if(url!=null &&(serverAvailable||adminMode)){
+          urlPreferenceChanged(mServerUrlPreference,url);
+          toastString +="URL : "+url+"\n";
+        }
+      }
+      catch (Exception e) {
+        toastString +="URL : not found\n";
+        Log.i(TAG,"Url not found");
+      }
+      try{
+        String username = (String) mainObject.get("username");
+        if(username!=null && (usernamePasswordAvailable || adminMode)){
+          onPreferenceChange(mUsernamePreference,username);
+         if(credentialAvailable || adminMode) signOnPreferenceChanged(mSignOnCredentialPreference,"username_password");
+          toastString = toastString +"Username : "+username+"\n";
+        }
+      }
+      catch (Exception e) {
+        toastString +="Sign-on Credential type: anonymous\n";
+        toastString +="Username : not found\n";
+        if(credentialAvailable || adminMode) signOnPreferenceChanged(mSignOnCredentialPreference,"none");
+        Log.i(TAG,"Username not found");
+      }
+
+     try{
+       String password = (String) mainObject.get("password");
+       if(password!=null && (usernamePasswordAvailable || adminMode)){
+         updatePassword(password);
+         toastString +="Password : "+password;
+       }
+     }
+     catch (Exception e) {
+       toastString +="Password : not found";
+       Log.i(TAG,"Password not found");
+     }
+
+     Toast.makeText(getActivity(), toastString, Toast.LENGTH_SHORT).show();
+
+    } catch (JSONException e) {
+      Toast.makeText(getActivity(), "Invalid Qr code.", Toast.LENGTH_SHORT).show();
+      Log.i(TAG,"Invalid QR code");
+    }
+
+  }
+
+  public void updatePassword(String pw) {
+    updatePreferenceSummary(CommonToolProperties.KEY_PASSWORD, pw);
+ }
+
+  @Override
+  public boolean onOptionsItemSelected(MenuItem item) {
+    switch (item.getItemId()) {
+      case R.id.action_barcode:
+        // When Scan QR icon is clicked.
+        if (ActivityCompat.checkSelfPermission(getActivity(), Manifest.permission.CAMERA)
+                == PackageManager.PERMISSION_GRANTED) {
+          // Permission is already available, start camera preview
+           openBarcodeScanner();
+        } else {
+          // Permission is missing and must be requested.
+          requestCameraPermission();
+        }
+        return true;
+
+      default:
+        return super.onOptionsItemSelected(item);
+
+    }
+  }
+
+  private void openBarcodeScanner() {
+    FragmentIntentIntegrator integrator = new FragmentIntentIntegrator(this);
```

Comment:
We can remove the FragmentIntentIntegrator class as you can chain the function calls together and use the factory method.  See the example below:

    IntentIntegrator.forFragment(this).
        setDesiredBarcodeFormats(IntentIntegrator.QR_CODE_TYPES).
        setPrompt("Place QR Code inside the rectangle").
        setCameraId(0).setBeepEnabled(true).setBarcodeImageEnabled(false).initiateScan();

Reply:
If you see any issues with this or any other comment, don't hesitate to let me know.  Thanks!

Reply:
That would work fine I guess. I do the changes.

Reply:
@clarlars,  I have done the required changes. Please review.